src/coreclr/tools/Common/TypeSystem/Ecma/EcmaType.TypeEquivalence.cs

```diff
@@ -0,0 +1,145 @@
+
+            string scope = (string)guidAttribute.Value.FixedArguments[0].Value;
+            string name = this.Name;
+            if (this.Namespace != null)
+                name = this.Namespace + "." + name;
+
+            return new TypeIdentifierData(scope, name);
+        }
+
+        private TypeIdentifierData ComputeTypeIdentifierData()
+        {
+            if (!Context.SupportsTypeEquivalence)
+                return null;
+
+            // Check for type identifier attribute
+            var typeIdentifierAttribute = this.GetDecodedCustomAttribute("System.Runtime.InteropServices", "TypeIdentifierAttribute");
+            if (typeIdentifierAttribute.HasValue)
+            {
+                // If the type has a type identifier attribute it is always considered to be type equivalent
+                if (typeIdentifierAttribute.Value.FixedArguments.Length == 0)
+                    return ComputeTypeIdentifierFromGuids();
+
+                if (typeIdentifierAttribute.Value.FixedArguments.Length != 2)
+                    return null;
+
+                if (typeIdentifierAttribute.Value.FixedArguments[0].Type != Context.GetWellKnownType(WellKnownType.String))
+                    return null;
+
+                if (typeIdentifierAttribute.Value.FixedArguments[1].Type != Context.GetWellKnownType(WellKnownType.String))
+                    return null;
+
+                _data = new TypeIdentifierData((string)typeIdentifierAttribute.Value.FixedArguments[0].Value, (string)typeIdentifierAttribute.Value.FixedArguments[1].Value);
+                return _data;
+            }
+            else
+            {
+                // In addition to the TypeIdentifierAttribute certain other types may also be opted in to type equivalence
+                if (Context.SupportsCOMInterop)
+                {
+                    // 1. Type is within assembly marked with ImportedFromTypeLibAttribute or PrimaryInteropAssemblyAttribute
+                    if (this.HasCustomAttribute("System.Runtime.InteropServices", "ImportedFromTypeLibAttribute") || this.HasCustomAttribute("System.Runtime.InteropServices", "PrimaryInteropAssemblyAttribute"))
+                    {
+                        // This type has a TypeIdentifier attribute if it has an appropriate shape to be considered type equivalent
+                    }
+
+                    if (!TypeHasCharacteristicsRequiredToBeTypeEquivalent)
+                        return null;
+
+                    _data = ComputeTypeIdentifierFromGuids();
+                }
+
+                return null;
+            }
+        }
+
+        public override TypeIdentifierData TypeIdentifierData
+        {
+            get
+            {
+                if (_data != null)
+                {
+                    if (object.ReferenceEquals(_data, TypeIdentifierData.Empty))
+                        return null;
+                    return _data;
+                }
+                lock(this)
```

Comment:
We don't generally lock on type system objects because with enough unchecked locks we'll eventually get deadlocks (see another spot where I'm policing this: https://github.com/dotnet/runtime/pull/87532#discussion_r1228848088). Would CompareExchange be unacceptable?

Reply:
Yes. This was just a bit of laziness on my part.